File: scoring.py
"""Scoring configuration and calculation.

Score = SCORE_TABLE[role][winner_side][is_alive] + adjustment
Adjustment captures moderator-entered bonuses (+) and penalties (−).
"""

from roles import ROLE_SIDES

# SCORE_TABLE[role][winner_side][is_alive: bool]
SCORE_TABLE: dict[str, dict[str, dict[bool, float]]] = {
    "citizen":    {"citizens": {True: 1.0, False: 0.5}, "mafia": {True: 0.0, False: 0.0}, "maniac": {True: 0.0, False: 0.0}},
    "sheriff":    {"citizens": {True: 2.0, False: 1.0}, "mafia": {True: 0.0, False: 0.0}, "maniac": {True: 0.0, False: 0.0}},
    "doctor":     {"citizens": {True: 2.0, False: 1.0}, "mafia": {True: 0.0, False: 0.0}, "maniac": {True: 0.0, False: 0.0}},
    "beauty":     {"citizens": {True: 2.0, False: 1.0}, "mafia": {True: 0.0, False: 0.0}, "maniac": {True: 0.0, False: 0.0}},
    "bodyguard":  {"citizens": {True: 1.0, False: 1.0}, "mafia": {True: 0.0, False: 0.0}, "maniac": {True: 0.0, False: 0.0}},
    "prosecutor": {"citizens": {True: 1.0, False: 0.5}, "mafia": {True: 0.0, False: 0.0}, "maniac": {True: 0.0, False: 0.0}},
    "con_artist": {"citizens": {True: 1.0, False: 0.5}, "mafia": {True: 0.0, False: 0.0}, "maniac": {True: 0.0, False: 0.0}},
    "thief":      {"citizens": {True: 1.0, False: 0.5}, "mafia": {True: 0.0, False: 0.0}, "maniac": {True: 0.0, False: 0.0}},
    "mafia":      {"citizens": {True: 0.0, False: 0.0}, "mafia": {True: 2.0, False: 1.5}, "maniac": {True: 0.0, False: 0.0}},
    "don":        {"citizens": {True: 0.0, False: 0.0}, "mafia": {True: 2.0, False: 1.5}, "maniac": {True: 0.0, False: 0.0}},
    "judge":      {"citizens": {True: 0.0, False: 0.0}, "mafia": {True: 2.0, False: 1.0}, "maniac": {True: 0.0, False: 0.0}},
    "maniac":     {"citizens": {True: 0.0, False: 0.0}, "mafia": {True: 0.0, False: 0.0}, "maniac": {True: 2.0, False: 2.0}},
}


def did_win(role: str, winner_side: str) -> bool:
    return ROLE_SIDES[role] == winner_side


def compute_score(role: str, winner_side: str, is_alive: bool, adjustment: float = 0.0) -> float:
    base = SCORE_TABLE[role][winner_side][is_alive]
    return round(base + adjustment, 2)

File: electron-app/src/utils/scoring.ts
// role → winner_side → alive → base score
export const SCORE_TABLE: Record<string, Record<string, Record<string, number>>> = {
  citizen: {
    citizens: { alive: 1.0, dead: 0.5 },
    mafia:    { alive: 0.0, dead: 0.0 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  sheriff: {
    citizens: { alive: 2.0, dead: 1.0 },
    mafia:    { alive: 0.0, dead: 0.0 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  doctor: {
    citizens: { alive: 2.0, dead: 1.0 },
    mafia:    { alive: 0.0, dead: 0.0 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  beauty: {
    citizens: { alive: 1.0, dead: 0.5 },
    mafia:    { alive: 0.0, dead: 0.0 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  bodyguard: {
    citizens: { alive: 1.0, dead: 1.0 },
    mafia:    { alive: 0.0, dead: 0.0 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  prosecutor: {
    citizens: { alive: 1.0, dead: 0.5 },
    mafia:    { alive: 0.0, dead: 0.0 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  con_artist: {
    citizens: { alive: 1.0, dead: 0.5 },
    mafia:    { alive: 0.0, dead: 0.0 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  thief: {
    citizens: { alive: 1.0, dead: 0.5 },
    mafia:    { alive: 0.0, dead: 0.0 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  mafia: {
    citizens: { alive: 0.0, dead: 0.0 },
    mafia:    { alive: 2.0, dead: 1.5 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  don: {
    citizens: { alive: 0.0, dead: 0.0 },
    mafia:    { alive: 2.0, dead: 1.5 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  judge: {
    citizens: { alive: 0.0, dead: 0.0 },
    mafia:    { alive: 2.0, dead: 1.0 },
    maniac:   { alive: 0.0, dead: 0.0 },
  },
  maniac: {
    citizens: { alive: 0.0, dead: 0.0 },
    mafia:    { alive: 0.0, dead: 0.0 },
    maniac:   { alive: 2.0, dead: 2.0 },
  },
}

export function computeScore(
  role: string,
  winnerSide: string,
  isAlive: boolean,
  adjustment = 0
): number {
  const aliveKey = isAlive ? 'alive' : 'dead'
  const base = SCORE_TABLE[role]?.[winnerSide]?.[aliveKey] ?? 0
  return Math.round((base + adjustment) * 100) / 100
}

export function didWin(role: string, winnerSide: string): boolean {
  const base = SCORE_TABLE[role]?.[winnerSide]?.['alive'] ?? 0
  return base > 0
}
